transports/webrtc/src/tokio/substream.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use asynchronous_codec::Framed;
+use bytes::Bytes;
+use futures::{channel::oneshot, prelude::*, ready};
+use tokio_util::compat::Compat;
+use webrtc::data::data_channel::{DataChannel, PollDataChannel};
+
+use std::{
+    io,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use crate::message_proto::{message::Flag, Message};
+use crate::tokio::{
+    substream::drop_listener::GracefullyClosed,
+    substream::framed_dc::FramedDC,
+    substream::state::{Closing, State},
+};
+
+mod drop_listener;
+mod framed_dc;
+mod state;
+
+/// Maximum length of a message.
+///
+/// "As long as message interleaving is not supported, the sender SHOULD limit the maximum message
+/// size to 16 KB to avoid monopolization."
+/// Source: <https://www.rfc-editor.org/rfc/rfc8831#name-transferring-user-data-on-a>
+const MAX_MSG_LEN: usize = 16384; // 16kiB
+/// Length of varint, in bytes.
+const VARINT_LEN: usize = 2;
+/// Overhead of the protobuf encoding, in bytes.
+const PROTO_OVERHEAD: usize = 5;
+/// Maximum length of data, in bytes.
+const MAX_DATA_LEN: usize = MAX_MSG_LEN - VARINT_LEN - PROTO_OVERHEAD;
+
+pub use drop_listener::DropListener;
+
+/// A substream on top of a WebRTC data channel.
+///
+/// To be a proper libp2p substream, we need to implement [`AsyncRead`] and [`AsyncWrite`] as well
+/// as support a half-closed state which we do by framing messages in a protobuf envelope.
+pub struct Substream {
+    io: FramedDC,
```

Comment:
```suggestion
    io: FramedDc,
```

Nit pick, Rust convention would be to call it `FramedDc` like done for `Arc`.